### PR TITLE
added logic to handle semicolon in the query

### PR DESCRIPTION
**Pull request by Manis99803**
## Description
Currently if we have a semicolon in the query, Pinot throws exception. However, semicolon are used in the SQL query to mark the termination of the query, and if a semicolon is present in the Pinot query exception shouldn't be thrown.
This PR implements the fix for the same

Issue Link: https://github.com/apache/pinot/issues/7713

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 1]
```diff
@@ -96,8 +96,36 @@ private CalciteSqlParser() {
...
+      // Split the input query based on semicolon
+      String[] sqlSplit = sql.split(";");
+
+      // If only semicolons are present in the input query
+      if (sqlSplit.length == 0) {
+        new SqlCompilationException("Caught exception while parsing query: " + sql);
+      } else {
+        // After spliting take the string present at 0th index and then parse the SQL
+        sql = sqlSplit[0];
+      }
```
Comment (richardstartin) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:119]: `String.split` is a bit heavy for this. Instead can you do 

```java
int semiColonIndex = sql.indexOf(';');
if (semiColonIndex == 0) {
    throw new SqlCompilationException("SQL query: " + sql + " starts with a semicolon.");
} else if (semiColonIndex > 0 {
    return sql.substring(0, semiColonIndex);
}
return sql;
```

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 2]
```diff
@@ -96,8 +96,36 @@ private CalciteSqlParser() {
...
       Pattern.compile("option\\s*\\(([^\\)]+)\\)", Pattern.CASE_INSENSITIVE);
 
+  /**
+   * Checks for the presence of semicolon in the sql query and modifies the query accordingly
+   *
+   * @param sql sql query
+   * @return sql query without semicolons
+   *
+   */
+  private static String checkForSemicolonInTheQuery(String sql) {
```
Comment (richardstartin) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:106]: I think we should rename this method to something more abstract, e.g. `sanitize`.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 1 above)
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:119]: Also, rather than just take the first statement, it should maybe throw if there is sql after a semicolon, in case the user thinks that multiple statements can be executed.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 3]
```diff
@@ -2387,4 +2387,48 @@ private void testSupportedDistinctQuery(String query) {
...
+  public void testInvalidQueryWithSemicolon() {
+    Assert.expectThrows(SqlCompilationException.class,
+            () -> CalciteSqlParser.compileToPinotQuery(";"));
+
+    Assert.expectThrows(SqlCompilationException.class,
+            () -> CalciteSqlParser.compileToPinotQuery(";;;;"));
+
+    Assert.expectThrows(SqlCompilationException.class,
+            () -> CalciteSqlParser.compileToPinotQuery("SELECT col1, count(*) FROM foo GROUP BY ; col1"));
+  }
```
Comment (richardstartin) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2433]: I really appreciate seeing good test cases along with this contribution 👍🏻 

One test case which I think should fail because only one of the submitted queries will be executed, but won't, is this one:

```
SELECT col1, count(*) FROM foo GROUP BY col1; 
SELECT col2, count(*) FROM foo GROUP BY col2 ;
```

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 4]
```diff
@@ -2387,4 +2387,48 @@ private void testSupportedDistinctQuery(String query) {
...
+    Assert.assertEquals(pinotQuery.getSelectList().get(0).getIdentifier().getName(), "col1");
+    Assert.assertEquals(pinotQuery.getGroupByListSize(), 1);
+    Assert.assertEquals(pinotQuery.getGroupByList().get(0).getIdentifier().getName(), "col1");
+    Assert.assertEquals(pinotQuery.getGroupByList().get(0).getIdentifier().getName(), "col1");
+
+    // Check for Option SQL Query
+    sql = "SELECT col1, count(*) FROM foo group by col1 option(skipUpsert=true);";
+    pinotQuery = CalciteSqlParser.compileToPinotQuery(sql);
+    Assert.assertEquals(pinotQuery.getQueryOptionsSize(), 1);
+    Assert.assertTrue(pinotQuery.getQueryOptions().containsKey("skipUpsert"));
```
Comment (richardstartin) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:2420]: Can you also add a test case for a literal with a semicolon in it, it should pass.

e.g. 

```
select col1, count(*) from foo where col1 = 'x;y' option(skipUpsert=true);
```

**Comment by richardstartin** [comment]
Hi @Manis99803 this may not be the right approach because stripping semicolons in a context insensitive way will break when literals which contain semicolons.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 1 above)
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:119]: My suggestion is to change the logic to allow only _terminating_ semicolons (like [beam](https://beam.apache.org/documentation/dsls/sql/calcite/lexical-structure/#terminating-semicolons)) - so only modify the query if there is only whitespace after the semicolon or it is the last character:

```java
int semiColonIndex = sql.indexOf(';');
boolean stripSemiColon = semiColonIndex >= 0;
for (int i = semiColonIndex; i < sql.length() && stripSemiColon; I++) {
  stripSemiColon &= Character.isWhitespace(sql.charAt(i));
}
return stripSemiColon ? sql.substring(0, semiColonIndex) : sql;
```

This way terminating semicolons can be stripped, but rejecting semicolons which can be ambiguous is left to the parser which has the intelligence to discern properly.

**Comment by Manis99803** [comment]
Hi @richardstartin ,
Thanks for the feedback, I get why this is not the right approach. I will work on the improvement and update the PR

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 2 above)
Comment (Manis99803) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: changed the function name with the one mentioned

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 1 above)
Comment (Manis99803) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: Updated the logic, really appreciate the feedback

**Comment by richardstartin** [comment]
Looks good assuming the tests pass 👍🏻

**Review by richardstartin — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 5]
```diff
@@ -96,8 +96,31 @@ private CalciteSqlParser() {
...
+   *
+   */
+  private static String removeTerminatingSemicolon(String sql) {
+    // Check if the query has semicolon
+    int semiColonIndex = sql.lastIndexOf(';');
+    int sqlLength = sql.length();
+    boolean stripSemiColon = semiColonIndex >= 0;
+
+    // check if semicolons are followed by whitespaces, only then termination should be done
+    for (int i = semiColonIndex + 1; i < sqlLength && stripSemiColon; i++) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:113]: Let's trim the sql first before checking for the last semicolon. This can help remove the unnecessary whitespace from the query, and we can simply check the character at the last index after trimming

**Review by Jackie-Jiang — COMMENTED** [review]
Thanks for the fix and adding the test

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 5 above)
Comment (Manis99803) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java]: Updated the PR as per the review comment

**Review by xiangfu0 — APPROVED** [review]
